### PR TITLE
fix(useArrowKeyNav): Focus lost when item is removed

### PR DESCRIPTION
**Pull request by mdodgelooker**
If `useArrowKeyNav`'s currently focused element is removed from the DOM, it becomes impossible to get focus back into the list. But as long as the list itself hasn't been removed, we want focus to stay inside until the user moves it elsewhere.

React does not trigger `onBlur` in this scenario. Some browsers do trigger a `focusout` event, but not all. The only reliable way to catch it is using a `MutationObserver` that checks if an element being removed was the focused item.

### Requirements

Please check the following items are addressed in your pull request (or are not applicable)

- [x] a11y impact (FUTURE: Make aXe pass req'd for CI ✅)

#### Image snapshots (choose one)

  - [ ] yes
  - [x] not applicable

#### Documentation updated (choose one)

  - [ ] yes
  - [x] not applicable

### CODE REVIEW

File: packages/components/src/utils/useArrowKeyNav.ts [hunk 1]
```diff
@@ -74,7 +83,7 @@ export const useArrowKeyNav = <E extends HTMLElement = HTMLElement>({
   onKeyDown,
 }: UseArrowKeyNavProps<E>) => {
   const internalRef = useRef<E>(null)
-  const [focusedItem, setFocusedItem] = useState<HTMLElement | null>(null)
+  const focusedItemRef = useRef<HTMLElement>()
```
Comment (mdodgelooker) [inline packages/components/src/utils/useArrowKeyNav.ts:86]: Realized this doesn't need to be a state value. We only need the latest, and no need to trigger a re-render.

File: packages/components/src/utils/useArrowKeyNav.ts [hunk 2]
```diff
@@ -134,6 +154,26 @@ export const useArrowKeyNav = <E extends HTMLElement = HTMLElement>({
...
   }
 
+  useEffect(() => {
+    const element = internalRef.current
+
+    // Create an observer to check if the focused element is removed
+    const observer = new MutationObserver((mutationsList) => {
+      if (checkElementRemoved(mutationsList, focusedItemRef.current)) {
+        // If it was, start focus back at the beginning
+        placeInitialFocus()
```
Comment (kvrmd) [inline packages/components/src/utils/useArrowKeyNav.ts:164]: Jared and I were just talking about this yesterday about this Date preset menu.

image

Comment (mdodgelooker) [reply packages/components/src/utils/useArrowKeyNav.ts:164]: That's the one! Terrible experience for keyboard nav right now. 😣

File: packages/components/src/List/List.story.tsx [hunk 3]
```diff
@@ -72,6 +72,39 @@ LongList.parameters = {
...
+        <ListItem
+          onClick={() => setShowMore(true)}
+          description="To test keyboard nav"
+        >
+          Show More
+        </ListItem>
+      )}
+    </List>
+  )
+}
```
Comment (kvrmd) [inline packages/components/src/List/List.story.tsx:102]: In dashboards-next Date preset filter, the entire list actually changes when More is clicked. 1 item in particular is added in between previously visible options. See the "Last 28 days" option below. Could we add a case when the menu list is completely reorganized?

image

Maybe something like this?
```
return (
  <List>
    {showMore ? (
      <ListGroup>
        <ListItem>Cheddar</ListItem>
        <ListItem>Swiss</ListItem>
      </ListGroup>
      <ListGroup>
        <ListItem>Gouda</ListItem>
        <ListItem>American</ListItem>
      </ListGroup>
      <ListItem>Show less</ListItem>
    ) : (
      <ListItem>Cheddar</ListItem>
      <ListItem>Gouda</ListItem>
      <ListItem>Show more</ListItem>
  )
)
```

**Review by kvrmd — DISMISSED** [review]
There's still some small issues as discussed with @mdodgelooker . But this is an overall improvement so +1.

File: packages/components/src/utils/useArrowKeyNav.ts [hunk 4]
```diff
@@ -113,27 +121,59 @@ export const useArrowKeyNav = <E extends HTMLElement = HTMLElement>({
...
-        const toFocus = getNextFocus(1, internalRef.current)
-        if (toFocus) {
-          // No need to update focusedItem with this since it's the default
-          toFocus.focus()
-        }
+        // Otherwise focus the first item
+        placeInitialFocus()
       }
+    } else {
+      focusedItemRef.current = e.target
```
Comment (mdodgelooker) [inline packages/components/src/utils/useArrowKeyNav.ts:149]: I was updating this in 2 other spots before, under the assumption that the arrow key handlers included here were the only way to move focus on the items. But screen readers have their own method to move focus (ctrl+option+arrow-right, in the `List` examples), so this is the obvious spot to keep track of the currently focused item. Thanks @kvrmd for catching that!

**Review by kvrmd — APPROVED** [review]
LGTM. Thank you

**Pull request merged**